File: src/ui/tab_scope.c
#include "ui_internal.h"

#include <stdlib.h>

typedef struct UITabBarScope {
    struct UITabBarScope *previous;
    int count;
    int selected;
    int item_open;
} UITabBarScope;

static UITabBarScope *tab_bar_scope;

void
ui_tab_scope_finish_frame(void)
{
    if(tab_bar_scope != NULL)
        abort();
}

int
BeginTabBar(TabBarProps bar, int *selected_index)
{
    UITabBarScope *scope;
    int clicked;
    int selected;

    if(bar.tabs == NULL || bar.count <= 0 || bar.bounds.width <= 0 ||
       bar.bounds.height <= 0)
        return 0;

    selected = selected_index != NULL ? *selected_index : bar.selected_index;
    if(selected < 0 || selected >= bar.count)
        selected = 0;
    bar.selected_index = selected;
    clicked = IsWindowReady() ? TabBar(bar) : ui_tab_bar_keyboard_input(bar);
    if(clicked >= 0) {
        selected = clicked;
        if(selected_index != NULL)
            *selected_index = clicked;
    } else if(selected_index != NULL && *selected_index != selected) {
        *selected_index = selected;
    }

    scope = calloc(1, sizeof(*scope));
    if(scope == NULL)
        abort();
    scope->previous = tab_bar_scope;
    scope->count = bar.count;
    scope->selected = selected;
    tab_bar_scope = scope;
    return 1;
}

int
BeginTabItem(int index)
{
    if(tab_bar_scope == NULL || tab_bar_scope->item_open)
        abort();
    if(index < 0 || index >= tab_bar_scope->count ||
       index != tab_bar_scope->selected)
        return 0;
    tab_bar_scope->item_open = 1;
    return 1;
}

void
EndTabItem(void)
{
    if(tab_bar_scope == NULL || !tab_bar_scope->item_open)
        abort();
    tab_bar_scope->item_open = 0;
}

void
EndTabBar(void)
{
    UITabBarScope *scope = tab_bar_scope;

    if(scope == NULL || scope->item_open)
        abort();
    tab_bar_scope = scope->previous;
    free(scope);
}

File: go/kryon/tab_scope.go
package kryon

// BeginTabBar renders the canonical tab header and opens a scope in which
// BeginTabItem exposes only the selected tab's arbitrary child widgets.
func (r *runtime) BeginTabBar(props TabBarProps, selectedIndex *int32) bool {
	count := props.Count
	if count <= 0 || count > int32(len(props.Tabs)) {
		count = int32(len(props.Tabs))
	}
	if count <= 0 || props.Bounds.Width <= 0 || props.Bounds.Height <= 0 {
		return false
	}

	selected := props.SelectedIndex
	if selectedIndex != nil {
		selected = *selectedIndex
	}
	if selected < 0 || selected >= count {
		selected = 0
	}
	props.Count = count
	props.SelectedIndex = selected
	if clicked := r.TabBar(props); clicked >= 0 {
		selected = clicked
	}
	if selectedIndex != nil {
		*selectedIndex = selected
	}
	r.tabBarScopes = append(r.tabBarScopes, tabBarScope{
		count:    count,
		selected: selected,
	})
	return true
}

func (r *runtime) BeginTabItem(index int32) bool {
	if len(r.tabBarScopes) == 0 {
		panic("BeginTabItem without BeginTabBar")
	}
	scope := &r.tabBarScopes[len(r.tabBarScopes)-1]
	if scope.itemOpen {
		panic("BeginTabItem before EndTabItem")
	}
	if index < 0 || index >= scope.count || index != scope.selected {
		return false
	}
	scope.itemOpen = true
	return true
}

func (r *runtime) EndTabItem() {
	if len(r.tabBarScopes) == 0 ||
		!r.tabBarScopes[len(r.tabBarScopes)-1].itemOpen {
		panic("EndTabItem without selected BeginTabItem")
	}
	r.tabBarScopes[len(r.tabBarScopes)-1].itemOpen = false
}

func (r *runtime) EndTabBar() {
	if len(r.tabBarScopes) == 0 {
		panic("EndTabBar without BeginTabBar")
	}
	n := len(r.tabBarScopes)
	if r.tabBarScopes[n-1].itemOpen {
		panic("EndTabBar before EndTabItem")
	}
	r.tabBarScopes[n-1] = tabBarScope{}
	r.tabBarScopes = r.tabBarScopes[:n-1]
}
